src/front/glsl/builtins.rs

```diff
@@ -0,0 +1,1650 @@
+use super::{
+    ast::{FunctionDeclaration, FunctionKind, ParameterInfo, ParameterQualifier},
+    context::Context,
+    Error, ErrorKind, Parser, Result, SourceMetadata,
+};
+use crate::{
+    BinaryOperator, Block, Expression, Handle, ImageClass, ImageDimension, ImageQuery,
+    MathFunction, Module, RelationalFunction, SampleLevel, Span, Type, TypeInner, VectorSize,
+};
+
+/// Helper function, to create a function prototype for a builtin
+fn add_builtin(
+    module: &mut Module,
+    args: Vec<TypeInner>,
+    builtin: Builtins,
+) -> FunctionDeclaration {
+    let mut parameters = Vec::with_capacity(args.len());
+    let mut parameters_info = Vec::with_capacity(args.len());
+
+    for arg in args {
+        parameters.push(module.types.fetch_or_append(
+            Type {
+                name: None,
+                inner: arg,
+            },
+            Span::Unknown,
+        ));
+        parameters_info.push(ParameterInfo {
+            qualifier: ParameterQualifier::In,
+            depth: false,
+        });
+    }
+
+    FunctionDeclaration {
+        parameters,
+        parameters_info,
+        kind: FunctionKind::Macro(builtin),
+        defined: false,
+        void: false,
+    }
+}
+
+impl Parser {
+    /// Injects a builtin into the [`lookup_function`](Self::lookup_function)
+    ///
+    /// This is done to not add a large startup cost and not increase memory
+    /// usage if it isn't needed.
+    pub(crate) fn inject_builtin(&mut self, name: String) {
```

Comment:
this isn't used anywhere?

Reply:
It's used in `function_call`, `add_prototype` and `add_function` otherwise the tests wouldn't pass.